### DISCUSSION TITLE
Export / save the best model

### DISCUSSION THREAD

**Discussion by simon-delisle** [microsoft/FLAML#228] [Q&A]
How can I save the best model to reuse the automl.predict later without fitting again?

Thanks,

Simon

**Comment by sonichi** [comment]
One way to do it is to use pickle, as demonstrated in this [notebook](https://github.com/microsoft/FLAML/blob/main/notebook/flaml_automl.ipynb).
```
import pickle
with open('automl.pkl', 'wb') as f:
    pickle.dump(automl, f, pickle.HIGHEST_PROTOCOL)
```

**Reply by Shashi9503** [reply]
> One way to do it is to use pickle, as demonstrated in this [notebook](https://github.com/microsoft/FLAML/blob/main/notebook/flaml_automl.ipynb).
> 
> ```
> import pickle
> with open('automl.pkl', 'wb') as f:
>     pickle.dump(automl, f, pickle.HIGHEST_PROTOCOL)
> ```

This is for saving the model. But how can I load it and reuse it ??

**Reply by qingyun-wu** [reply]
You can use pickle.load() to load the automl instance you saved and then predict without fitting again.

Code example assuming your model is dumped into file 'automl.pkl' and test data is X_test:
```
automl = pickle.load(open('automl.pkl', 'rb'))
y_pred = automl.predict(X_test)
```

**Reply by Shashi9503** [reply]
this worked, thanks

**Comment by eavidan** [comment]
is it possible to export the model to its native package?
meaning, using `xgboost` to load the model rather then `flaml`
One application would be to continue working on the model with its native package

**Reply by sonichi** [reply]
Yes. `automl.model.estimator` is the native package's estimator. For example, if it's a xgboost estimator, you can do `automl.model.estimator.save_model(filename)`.

**Reply by eavidan** [reply]
Thanks @sonichi this is great. what about the preprocess?
I was looking at the code and it looks like each model implementation of `BaseEstimator` includes a _preprocess.
Is there any API I should you in order to invoke it, or use directly `automl.model._preprocess()`?

**Reply by sonichi** [reply]
There are two preprocessors: task-level and estimator-level. The former is `automl._preprocess()` and the latter is `automl.model._preprocess()`. The former should be called before the latter. Right now they are not exposed as user-facing APIs.
We can add some user-facing APIs in future version. If you have any ideas for that please feel free to suggest.

**Reply by eavidan** [reply]
thanks for the details @sonichi. I looked at the code and now I believe I understand better the flow of things.
my first thought is why not expose both as a user-facing APIs: `automl.preprocess` and `automl.model.preprocess()` just so we know these methods are hear to stay.
I agree that getting the entire preprocessing chained together through a single autonomous API would probably require some refactoring. will give it some thought and will try to suggest something.
for now, I will simply chain them together using `flaml`. thanks for all the info !

**Reply by sonichi** [reply]
Thanks. An issue has been created: https://github.com/microsoft/FLAML/issues/842. Let's make future discussions over there.
